core/lib/multivm/src/versions/vm_latest/types/l1_batch.rs

```diff
@@ -0,0 +1,41 @@
+use crate::interface::L1BatchEnv;
+use std::collections::HashMap;
+use zksync_types::U256;
+use zksync_utils::{address_to_u256, h256_to_u256};
+
+const OPERATOR_ADDRESS_SLOT: usize = 0;
+const PREV_BLOCK_HASH_SLOT: usize = 1;
+const NEW_BLOCK_TIMESTAMP_SLOT: usize = 2;
+const NEW_BLOCK_NUMBER_SLOT: usize = 3;
+const L1_GAS_PRICE_SLOT: usize = 4;
+const FAIR_L2_GAS_PRICE_SLOT: usize = 5;
+const EXPECTED_BASE_FEE_SLOT: usize = 6;
+const SHOULD_SET_NEW_BLOCK_SLOT: usize = 7;
+
+/// Returns the initial memory for the bootloader based on the current batch environment.
+pub(crate) fn bootloader_initial_memory(l1_batch: &L1BatchEnv) -> Vec<(usize, U256)> {
+    let mut base_params: HashMap<usize, U256> = vec![
+        (
+            OPERATOR_ADDRESS_SLOT,
+            address_to_u256(&l1_batch.fee_account),
+        ),
+        (PREV_BLOCK_HASH_SLOT, Default::default()),
+        (NEW_BLOCK_TIMESTAMP_SLOT, U256::from(l1_batch.timestamp)),
+        (NEW_BLOCK_NUMBER_SLOT, U256::from(l1_batch.number.0)),
+        (L1_GAS_PRICE_SLOT, U256::from(l1_batch.l1_gas_price)),
+        (
+            FAIR_L2_GAS_PRICE_SLOT,
+            U256::from(l1_batch.fair_l2_gas_price),
+        ),
+        (EXPECTED_BASE_FEE_SLOT, U256::from(l1_batch.base_fee())),
+        (SHOULD_SET_NEW_BLOCK_SLOT, U256::from(0u32)),
+    ]
+    .into_iter()
+    .collect();
+
+    if let Some(prev_block_hash) = l1_batch.previous_batch_hash {
+        base_params.insert(PREV_BLOCK_HASH_SLOT, h256_to_u256(prev_block_hash));
+        base_params.insert(SHOULD_SET_NEW_BLOCK_SLOT, U256::from(1u32));
+    }
+    base_params.into_iter().collect()
+}
```

Comment:
This is a bit heavy with first creating a vec then creating a `HashMap` and then mutating it back to the `vec`. you could do something like

```rust
let (prev_block_hash, should_set_new_block_slot) = l1_batch.previous_batch_hash.map(|h| (h256_to_u256(prev_block_hash), U256::from(1u32)).unwrap_or_default();
```

and then just plug these values into a `vec![]` initialization and just return this `vec![]` since the default value for `U256` is 0 and for prev block hash slot it is default already.